IdentityCore/src/cache/ios/MSIDKeychainTokenCache.m

```diff
@@ -54,7 +242,185 @@ - (BOOL)removeItemWithKey:(MSIDTokenCacheKey *)key
                                context:(id<MSIDRequestContext>)context
                                  error:(NSError **)error
 {
-    return nil;
+    assert(serializer);
+    
+    MSID_LOG_INFO(context, @"Get keychain items, key info (account: %@ service: %@)", _PII_NULLIFY(key.account), _PII_NULLIFY(key.service));
+    MSID_LOG_INFO_PII(context, @"Get keychain items, key info (account: %@ service: %@)", key.account, key.service);
+    
+    NSMutableDictionary *query = [self defaultQuery];
+    if (key.service)
+    {
+        [query setObject:key.service forKey:(id)kSecAttrService];
+    }
+    if (key.account)
+    {
+        [query setObject:key.account forKey:(id)kSecAttrAccount];
+    }
+    [query setObject:@YES forKey:(id)kSecReturnData];
+    [query setObject:@YES forKey:(id)kSecReturnAttributes];
+    [query setObject:(id)kSecMatchLimitAll forKey:(id)kSecMatchLimit];
+    
+    CFTypeRef cfItems = nil;
+    MSID_LOG_INFO(context, @"Trying to find keychain items...");
+    OSStatus status = SecItemCopyMatching((CFDictionaryRef)query, &cfItems);
+    MSID_LOG_INFO(context, @"Keychain find status: %d", status);
+    
+    if (status == errSecItemNotFound)
+    {
+        return @[];
+    }
+    else if (status != errSecSuccess)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(NSOSStatusErrorDomain, status, @"Failed to get items from keychain.", nil, nil, nil, context.correlationId, nil);
+        }
+        MSID_LOG_ERROR(context, @"Failed to find keychain item (status: %d)", status);
+        return nil;
+    }
+    
+    NSArray *items = CFBridgingRelease(cfItems);
+    
+    NSMutableArray *tokenItems = [[NSMutableArray<MSIDToken *> alloc] initWithCapacity:items.count];
+    for (NSDictionary *attrs in items)
+    {
+        NSData *itemData = [attrs objectForKey:(id)kSecValueData];
+        MSIDToken *tokenItem = [serializer deserialize:itemData];
+        
+        if (tokenItem)
+        {
+            // Delete tombstones generated from previous versions of ADAL.
+            if (tokenItem.tokenType == MSIDTokenTypeRefreshToken && [tokenItem.token isEqualToString:@"<tombstone>"])
+            {
+                [self deleteTombstoneWithService:attrs[(id)kSecAttrService]
+                                         account:attrs[(id)kSecAttrAccount]
+                                         context:context];
+            }
+            else
+            {
+                [tokenItems addObject:tokenItem];
+            }
+        }
+        else
+        {
+            MSID_LOG_ERROR(context, @"Failed to deserialize token item.");
```

Comment:
What would be the reason for deserialization to fail? Should we instead fail the whole call?

Reply:
Deserialization can fail if data was corrupted in keychain, but usually it should not happen. We should not fail the whole call because ADAL skip's items that failed to serialize.